rs/sns/governance/src/lib.rs

```diff
@@ -121,12 +126,35 @@ fn validate_required_field<'a, Inner>(
         .ok_or_else(|| format!("The {} field must be populated.", field_name))
 }
 
-/// Return an Err whose inner value describes (in detail) what is wrong with a
-/// field value, and where within some (Protocol Buffers message) struct.
-fn field_err(field_name: &str, field_value: impl Debug, defect: &str) -> Result<(), String> {
+/// Return an Err whose inner value describes (in detail) what is wrong with a field value (should
+/// be bounded), and where within some (Protocol Buffers message) struct.
+///
+/// Only up to the first `MAX_SCALAR_FIELD_LEN_BYTES` bytes will be taken from `field_value`.
+fn field_err(field_name: &str, field_value: String, defect: &str) -> Result<(), String> {
+    let mut bounded_field_value = String::new();
+    // Searching for the longest byte prefix that is a valid string is expensive in the worst case.
+    // This is because unicode characters are unbounded. Instead, concatenate characters one-by-one,
+    // until we either run out of characters or exceed the limit (in which case we pop the last
+    // character to still comply with the limits). Note that a `char` is always 4 bytes, but
+    // pushing it to a string does not always increase a string's byte size by 4 bytes.
+    // For example:
+    // ```
+    // println!("bytes = {}", std::mem::size_of_val(&'\u{200D}')); // bytes = 4
+    // println!("bytes = {}", std::mem::size_of_val("\u{200D}"));  // bytes = 3
+    // ```
+    for c in field_value.chars() {
+        bounded_field_value.push(c);
+        if bounded_field_value.len() > MAX_SCALAR_FIELD_LEN_BYTES {
+            bounded_field_value.pop();
+            break;
+        }
+    }
```

Comment:
That function doesn't seem safe for arbitrary unicode strings. For example, it would panic in the following case:

```
clamp_string_len("\u{200D}\u{200D}\u{200D}", 2)
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=16ace03d1cfbfb292b28b2e96b9cbe1f